Add JSitor liquid tag

artoodeeto opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

<!--- For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments -->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
Added a feature for jsitor liquid

## Related Tickets & Documents
Resolves #4214

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ ] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

artoodeeto commented:
I didn't include any images because its already in #4214 but if you want ill add it. Please let me know where can I improve more and also sorry for the delay had to bring home the work. 🙏

ashvin777 commented on app/liquid_tags/jsitor_tag.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  def jsitor_error
+    raise StandardError, "Invalid jsitor link"
```
Suggestion - If we can add specific error by adding the example link then it will be more awesome. :-)

rhymes replied:
also `JSitor` instead of `jsitor` :)

rhymes commented on app/liquid_tags/jsitor_tag.rb:
```diff
@@ -0,0 +1,38 @@
+class JsitorTag < LiquidTagBase
+  PARTIAL = "liquids/jsitor".freeze
+  JSITOR_URL = /\A(https|http):\/\/(jsitor)\.(com)\/(embed)\/([a-zA-Z0-9]+)\Z/.freeze
+  def initialize(tag_name, link, token)
```
you could add a space here

rhymes commented on app/liquid_tags/jsitor_tag.rb:
```diff
@@ -0,0 +1,38 @@
+class JsitorTag < LiquidTagBase
+  PARTIAL = "liquids/jsitor".freeze
+  JSITOR_URL = /\A(https|http):\/\/(jsitor)\.(com)\/(embed)\/([a-zA-Z0-9]+)\Z/.freeze
```
`URL_REGEXP` seems a better name :)

rhymes commented on app/views/pages/_editor_guide_text.html.erb:
```diff
@@ -328,6 +328,9 @@
         <h3><strong>Asciinema Embed</strong></h3>
         <p>All you need is the Asciinema id:</p>
         <code>{% asciinema 239367 %}</code>
+        <h3><strong>Jsitor Liquid Tag</strong></h3>
```
`JSitor`

rhymes commented on app/views/pages/_editor_guide_text.html.erb:
```diff
@@ -328,6 +328,9 @@
         <p>All you need is the Asciinema id:</p>
         <code>{% asciinema 239367 %}</code>
+        <h3><strong>Jsitor Liquid Tag</strong></h3>
+        <p>Copy jsitor embeddable link make sure no slash (/) on the end of the <code>link</code>, it should have the <em>embed</em> path, and ends with ID code, as follows:</p>
```
`JSitor`

artoodeeto commented:
@ashvin777 @rhymes hello guys. thanks for the comments. I will add theses changes tonight. also my bad for the branch name. thought it was alright. :) thanks again

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=4657) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=4657) before we can accept your contribution.<br/>**13** out of **14** committers have signed the CLA.<br/><br/>:white_check_mark: artoodeeto<br/>:white_check_mark: AndyStabler<br/>:white_check_mark: mstruve<br/>:white_check_mark: juliannatetreault<br/>:white_check_mark: rhymes<br/>:white_check_mark: bradpauly<br/>:white_check_mark: hydrosquall<br/>:white_check_mark: ikanyu<br/>:white_check_mark: obahareth<br/>:white_check_mark: jennapederson<br/>:white_check_mark: garethx<br/>:white_check_mark: 3b3ziz<br/>:white_check_mark: danhuynhdev<br/>:x: dependabot-preview[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=4657) it.</sub>

artoodeeto commented:
Jesus Christ Sorry I made a mistake on rebasing. 😢

artoodeeto commented:
@rhymes sir I messed up. I updated my origin branch. what I did was on my master branch:

1. git pull upstream master
2. git rebase upstream/master
3. git push origin master
4. checked out this feature branch
5. git rebase master
6. git push origin/feature-branch

and then there were commits that unrelated to this branch so I googled and followed [this](https://stackoverflow.com/a/51400593/7034540). 

basically, to fix it I dropped the 23 unrelated commits:
1. git rebase -i HEADE~23
2. drop the 23
3. save and exit
4. git push --force-with-lease

but I dont know if it will affect the history. im really sorry. also on my last commit there's a big green maybe thats on the space between regex and initialize method? my bad. sorry. next time ill be very careful and if I get ban I understand.

rhymes commented:
@artoodeeto don't worry, git mistakes happen! the great thing is that they are usually fixable. i'll get another review in

artoodeeto commented:
@rhymes I'm not really sure how to test it now. removed a whole lot since I just need to accept the id then recreate the link. I don't have really to check the id that is being passed because it will be parse as a string anyway. what do you think?

rhymes commented on app/views/liquids/_jsitor.html.erb:
```diff
@@ -0,0 +1,8 @@
+  scrolling="no"
+  frameborder="no"
+  allowtransparency="true"
+  style="width: 100%;">
```
could you add `loading="lazy"` to this? Since there's a PR going around adding lazy loading to iframes, we might as well get that in right away here :D

artoodeeto replied:
ohh. hey sir is that the right behavior? @ashvin777. 

@rhymes ill add it when I get back home :smile: also are my test suite fine? I dont know how to test that kind of behavior, but in Integration test maybe?

rhymes replied:
@artoodeeto I think we should be fine without testing the "defaultness" of jsitor when the id is incorrect

ashvin777 replied:
Thank @rhymes for noticing this issue. I will fix this behavior. Ideally, it should show a "404" error page.

rhymes replied:
I think the current behavior is fine as well, I'll let you evaulate how hard it'd be to intercept something like that

artoodeeto replied:
@ashvin777 @rhymes we could check the response status? make an HTTP request on the backend check status if 2** then continue loading the page 4** then raise an error? what do you think?

rhymes replied:
Don't think so.

```shell
➜  ~ curl -I https://jsitor.com/foobar
HTTP/1.1 200 OK
➜  ~ curl -I https://jsitor.com/embed/foobar
HTTP/1.1 200 OK
```

I'm thinking we shoulnd't get stuck on this detail, it likely won't matter much because the publisher will likely check if their embed is correct before publishing.

As long as there are no security fallacies here I don't think spending time to intercept if the ID is correct or not is worth it.

artoodeeto replied:
@rhymes  I agree with you. also we have the guide to "guide" them lols. but we could technically do it if @ashvin777 will update the site to return 400.

artoodeeto replied:
heres my fix for lazy loadig. 😄 @rhymes

ashvin777 commented:
Hi @rhymes and @artoodeeto , is it possible to convert the regular link `https://jsitor.com/B7FQ5tHbY` into embed link `https://jsitor.com/embed/B7FQ5tHbY`  internally ?

So if user enters the embed template like below 
`${ jsitor https://jsitor.com/B7FQ5tHbY }`

It should load `https://jsitor.com/embed/B7FQ5tHbY` link instead?

rhymes commented:
@ashvin777 good idea

artoodeeto commented:
> Hi @rhymes and @artoodeeto , is it possible to convert the regular link `https://jsitor.com/B7FQ5tHbY` into embed link `https://jsitor.com/embed/B7FQ5tHbY` internally ?
> 
> So if user enters the embed template like below
> `${ jsitor https://jsitor.com/B7FQ5tHbY }`
> 
> It should load `https://jsitor.com/embed/B7FQ5tHbY` link instead?

@rhymes @ashvin777  but we already have the guide though? if they check the guide they will read that it's just the id is being required to load it. no?

rhymes commented:
@artoodeeto yeah, I think it can be as is right now, but there are liquid tags like the listings tag that supports both the "id" and the URL :)

If you add the lazy loading bit I think the PR can be accepted as is, bar a second set of eyes in case I've missed something

artoodeeto commented:
@ashvin777 @rhymes welcome welcome. :smile:

ashvin777 commented:
Hi @artoodeeto , Thanks you so much for this work. 

I just saw the production guide (dev.to) of JSitor published after merging this PR

![image](https://user-images.githubusercontent.com/8143154/68222711-72483400-0011-11ea-8646-8ac3ba9be8ed.png)

Which seems incorrect, The request was to use the URL instead of the id. The reason behind that is, there are some additional query params which can be used to hide/show some of the tabs in embeddable content, for example, to see the JS tab content only then we can do it using https://jsitor.com/embed/1QgJVmCam?js

A light theme embeddable -
https://jsitor.com/embed/1QgJVmCam?html&js&css&result&light&

All of these options are configurable in embed section of JSitor
![image](https://user-images.githubusercontent.com/8143154/68223126-35c90800-0012-11ea-9e11-04b1140bffd5.png)

Is it possible to either change this to use URL or else is there any other way of passing this query params from this liquid tag?

Zhao-Andy commented:
Hi @ashvin777, thanks for raising that. I missed your comment earlier about it before merging it. I think having the current iteration is a good start, and we can add the ability to handle query parameters as well as use the full URL easily.

@artoodeeto if you'd like to continue with this and work on that, you're more than welcome to! Otherwise I'll make an issue and leave the task up for grabs.

artoodeeto commented:
@ashvin777 sorry about that, I thought we only need the embed with id so I didn't bother about the query params. ill redo this. :smile: probably this weekend.

@Zhao-Andy can you make an issue on this one and assign it to me? I don't want to use the current branch because it's broken. If I make a push its always asking me to pull first so I have to do a force push. if it's ok with you. :)

@rhymes hello sir just wanted to tag you. haha. :v:

rhymes commented:
@artoodeeto yeah, you shouldn't use a merged branch, you can delete it and branch off from a fresh master, don't forget to update your copy of master with `git checkout master && git pull upstream master`

artoodeeto commented:
@rhymes yeap. Ill im gonna grab the new issue. thank you! :heart:
